### PR TITLE
HIVE-26233: Problems reading back PARQUET timestamps above 10000 years

### PR DESCRIPTION
**Pull request by pvary**
### What changes were proposed in this pull request?
Timestamps where the year is more than 4 digits will not have an extra + sign in front of them

### Why are the changes needed?
So these big timestamps could be read and rewritten so they can conform the original 9999 year limit.

### Does this PR introduce _any_ user-facing change?
Timestamps where the year is more than 4 digits will not have an extra + sign in front of them. Since it was not supported even before this change, I think it is ok to have this in.

### How was this patch tested?
Added unit test

### CODE REVIEW

File: ql/src/test/org/apache/hadoop/hive/ql/io/parquet/serde/TestParquetTimestampUtils.java [hunk 1]
```diff
@@ -389,6 +390,26 @@ public void testIllegalInt64TimestampStrings() {
     verifyInt64TimestampValue("2262-04-11 23:47:16.854775808", LogicalTypeAnnotation.TimeUnit.NANOS, false);
   }
 
+  @Test
```
Comment (zabetak) [inline ql/src/test/org/apache/hadoop/hive/ql/io/parquet/serde/TestParquetTimestampUtils.java:393]: The test passes successfully if you exclude all the changes in this PR and revert some lines in `NanoTimeUtils` as I mentioned in the JIRA case. This makes me think that we need to treat the problem differently since it does not appear to be related to the formatter.

**Review by zabetak — CHANGES_REQUESTED** [review]
This appears to be a regression caused by HIVE-22589. I left some comments under HIVE-26233 so let's continue the discussion there.

File: ql/src/test/org/apache/hadoop/hive/ql/io/parquet/serde/TestParquetTimestampUtils.java (hunk 1 above)
Comment (zabetak) [reply ql/src/test/org/apache/hadoop/hive/ql/io/parquet/serde/TestParquetTimestampUtils.java:393]: What I wrote above is not true. Reverting the changes about the proleptic calendar in `NanoTimeUtils` does not fix this test. I don't know why the test was passing before but I suspect a "stale" workspace.

File: common/src/java/org/apache/hadoop/hive/common/type/Timestamp.java [hunk 2]
```diff
@@ -101,6 +101,16 @@ public class Timestamp implements Comparable<Timestamp> {
...
+  private static final DateTimeFormatter PRINT_LENIENT_FORMATTER = new DateTimeFormatterBuilder()
+      // Date and Time Parts
+      .appendValue(YEAR, 4, 10, SignStyle.NORMAL).appendLiteral('-').appendValue(MONTH_OF_YEAR, 2, 2, SignStyle.NORMAL)
+      .appendLiteral('-').appendValue(DAY_OF_MONTH, 2, 2, SignStyle.NORMAL)
+      .appendLiteral(" ").appendValue(HOUR_OF_DAY, 2, 2, SignStyle.NORMAL).appendLiteral(':')
+      .appendValue(MINUTE_OF_HOUR, 2, 2, SignStyle.NORMAL).appendLiteral(':')
+      .appendValue(SECOND_OF_MINUTE, 2, 2, SignStyle.NORMAL)
+      // Fractional Part (Optional)
+      .optionalStart().appendFraction(ChronoField.NANO_OF_SECOND, 0, 9, true).optionalEnd().toFormatter();
+
```
[the remark this answers is not in the document]
Comment (zabetak) [reply common/src/java/org/apache/hadoop/hive/common/type/Timestamp.java:113]: Also using `LENIENT` in the name is a bit misleading since it implies that `DateTimeFormatterBuilder#parseLenient` is in use which is not the case here.

File: common/src/java/org/apache/hadoop/hive/common/type/Timestamp.java [hunk 3]
```diff
@@ -128,6 +138,10 @@ public String toString() {
     return localDateTime.format(PRINT_FORMATTER);
   }
 
+  public String toStingWithLenientFormatter() {
+    return localDateTime.format(PRINT_LENIENT_FORMATTER);
+  }
+
```
Comment (zabetak) [inline common/src/java/org/apache/hadoop/hive/common/type/Timestamp.java:144]: The use of `Lenient` is a bit misleading as I wrote previously. Also there is a small typo in the method name `toSting` vs `toString`.

Instead of adding a new method we could use `Timestamp#format` passing in the desired formatter. With the right naming for the formatter parameter it would make the intention more clear.

File: ql/src/test/org/apache/hadoop/hive/ql/io/parquet/serde/TestParquetTimestampUtils.java [hunk 4]
```diff
@@ -389,6 +390,26 @@ public void testIllegalInt64TimestampStrings() {
...
+  public void testTimestamp9999() {
```
Comment (zabetak) [inline ql/src/test/org/apache/hadoop/hive/ql/io/parquet/serde/TestParquetTimestampUtils.java:394]: What do you think of refactoring and moving the test in `TestParquetTimestampsHive2Compatibility` which exactly about compatibility with Hive 2?

**Review by zabetak — APPROVED** [review]
LGTM! Left a few comments but nothing major; feel free to disregard.

File: common/src/java/org/apache/hadoop/hive/common/type/Timestamp.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (pvary) [reply common/src/java/org/apache/hadoop/hive/common/type/Timestamp.java]: Moved to TimestampTZUtil, and renamed

File: ql/src/test/org/apache/hadoop/hive/ql/io/parquet/serde/TestParquetTimestampUtils.java (hunk 4 above)
Comment (pvary) [reply ql/src/test/org/apache/hadoop/hive/ql/io/parquet/serde/TestParquetTimestampUtils.java:397]: I think for a single tests we should not create a separate test class.
Added comment to make clear what we are testing here.

File: common/src/java/org/apache/hadoop/hive/common/type/Timestamp.java (hunk 3 above)
Comment (pvary) [reply common/src/java/org/apache/hadoop/hive/common/type/Timestamp.java]: Good idea with the `format()` method

File: ql/src/test/org/apache/hadoop/hive/ql/io/parquet/serde/TestParquetTimestampUtils.java (hunk 4 above)
Comment (zabetak) [reply ql/src/test/org/apache/hadoop/hive/ql/io/parquet/serde/TestParquetTimestampUtils.java:397]: I was thinking to move it inside the existing class, not create a new one.

Comment (zabetak) [reply ql/src/test/org/apache/hadoop/hive/ql/io/parquet/serde/TestParquetTimestampUtils.java:397]: Or rather enrich the tests inside `TestParquetTimestampsHive2Compatibility` to account for this edge case.

Comment (pvary) [reply ql/src/test/org/apache/hadoop/hive/ql/io/parquet/serde/TestParquetTimestampUtils.java:397]: Th ave never seen `TestParquetTimestampsHive2Compatibility`, so it is good that you highlighted here. OTOH that is a parameterized test, without TZ, so I think it would be a full rewrite to get this work for this specific TZ and TS

Comment (zabetak) [reply ql/src/test/org/apache/hadoop/hive/ql/io/parquet/serde/TestParquetTimestampUtils.java:397]: How about something like the following:

```java
  private static Stream<String> generateTimestamps() {
    return Stream.concat(Stream.of("9999-12-31 23:59:59.999"), Stream.generate(new Supplier<String>() {
```

Comment (zabetak) [reply ql/src/test/org/apache/hadoop/hive/ql/io/parquet/serde/TestParquetTimestampUtils.java:397]: `TestParquetTimestampsHive2Compatibility` also uses multiple timezones so I think it should be fine and adding another one should be trivial.

File: ql/src/test/org/apache/hadoop/hive/ql/io/parquet/serde/TestParquetTimestampsHive2Compatibility.java [hunk 5]
```diff
@@ -79,6 +79,18 @@ void testWriteHive2ReadHive4UsingLegacyConversion(String timestampString) {
...
   }
 
+  /**
+   * Tests that timestamps written using Hive2 APIs are read correctly by Hive4 APIs when legacy conversion is on.
+   */
+  @ParameterizedTest(name = "{0}")
+  @MethodSource("generateTimestamps")
+  void testWriteHive2ReadHive4UsingLegacyConversionWithZone(String timestampString) {
+    String zoneId = "US/Pacific";
+    NanoTime nt = writeHive2(timestampString);
```
Comment (zabetak) [inline ql/src/test/org/apache/hadoop/hive/ql/io/parquet/serde/TestParquetTimestampsHive2Compatibility.java:89]: Since there is no parameter in `writeHive2` for specifying the timezone I think you will need to call `TimeZone.setDefault()` explicitly otherwise it will not work.
